### DISCUSSION TITLE
shadowDatabaseUrl in schema.prisma appears to be ignored by the cli

### DISCUSSION THREAD

**Discussion by smarchesini-DQ** [prisma/prisma#20414] [Q&A] [answered]
Hi I am using prisma inside redwoodjs, and I am using the shadow database for reconciling db drifts. 
Everything works great when I pass the shadow-database-url to the migrate diff command.

So I wanted to make the shadow db url configurable, and following the info from https://github.com/prisma/prisma/releases/tag/2.17.0 I set

```
datasource db {
  provider          = "postgresql"
  url               = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
}
```
still if I do not pass the --shadow-database-url to the cli call I get an error

option requires argument: --shadow-database-url

I double checked that the env var is present in ${process.env.SHADOW_DATABASE_URL}

Any hint on how to debug this?
TIA

**Comment by janpio** [comment]
`migrate diff` by default does not use the value from the `schema.prisma` I think, but always requires the parameter to be explicitly set.
It has some parameters that read the schema file and use the models from there or similar, but I don't think that applies to the shadow database URL.

What command exactly are you running where you would expect it to be used?

**Reply by smarchesini-DQ** [reply]
Thanks for your reply.  I use

`prisma migrate diff --from-migrations api/db/migrations --to-schema-datamodel api/db/schema.prisma --shadow-database-url <myShadowdbUrl> --script`

which works perfectly for my purposes, it would be just useful for us to be able to configure the shadow db url, so I was looking at
https://github.com/prisma/prisma/issues/4751
and found about setting the url in the schema datasource.

I guess this is not read by the cli command? I guess I can just create a script for that, but now I am wondering, what is the use case for that setting in the schema datasource then?

thank you

**Reply by janpio** [reply, marked as answer]
`migrate dev` uses that value.
`migrate diff` unfortunately does not yet.
You could create a feature request for this (if one does not even exist yet).

**Comment by smarchesini-DQ** [comment]
Thank you so much, I'd love to contribute a PR but for the life of me I can not find in the source code where the migrate command accesses shadowDatabaseUrl from the schema, I searched for 'shadow' all over and didn't find it, now I am so curious 🐙

**Reply by janpio** [reply]
The `migrate dev` logic will not be in `prisma/prisma`, but the `prisma-engines` repository. The CLI hands over control of executing the logic to our Schema Engine, which runs all Migration and Introspection-related things. For `migrate diff`, that code is probably missing or disabled (or just not relavant as it does things differently).

**Reply by janpio** [reply]
Some superficial sleuthing:
- This is the CLI code of `diff` that collects all the parameters and then hands over to the implementation of that engine: https://github.com/prisma/prisma/blob/main/packages/migrate/src/commands/MigrateDiff.ts
- The relevant implementation bit is rather small: https://github.com/prisma/prisma/blob/7854a5f521278ac0abd53c9386f29c2da8d6a0fc/packages/migrate/src/SchemaEngine.ts#L209-L211
- And this is part of the actual implementation (I could not easily find where this would enter the engine, but this seems to be core of the implementation): https://github.com/prisma/prisma-engines/blob/main/schema-engine/core/src/commands/diff.rs

Somewhere in there it would need to automatically try to load the `shadowDatabaseUrl` property of the schema file instead of only relying on the provided parameter. My own abilities end here as well, no idea where and how that would need to happen.

**Comment by smarchesini-DQ** [comment]
Thank you, just for reference this is the commit where it was made available for migrate diff. (thx for pointing me to the engine repos, now I have a better understanding of prisma architecture)
https://github.com/prisma/prisma-engines/commit/df569f7e828d745b2ce50108f5236cdbe863f76c

**Reply by janpio** [reply]
That is a good find! 

I think though, that this made it available everywhere _but_ `migrate diff`, or that `migrate diff` was actually added later and uses a different method to load the shadow database information. Still, understanding this is a good first step to then also understand why this does not apply to `migrate diff`.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
